File: app/src/main/java/com/petal/browser/browser/BrowserContainer.java
package com.petal.browser.browser;

import java.util.LinkedList;
import java.util.List;

import com.petal.browser.view.NinjaWebView;

public class BrowserContainer {
    private static final List<AlbumController> list = new LinkedList<>();

    public static AlbumController get(int index) {
        return list.get(index);
    }

    public synchronized static void add(AlbumController controller) {
        list.add(controller);
    }

    public synchronized static void add(AlbumController controller, int index) {
        list.add(index, controller);
    }

    public synchronized static void remove(AlbumController controller) {
        if (controller instanceof NinjaWebView) {
            ((NinjaWebView) controller).destroy();
        }
        list.remove(controller);
    }

    public static int indexOf(AlbumController controller) {
        return list.indexOf(controller);
    }

    public static List<AlbumController> list() {
        return list;
    }

    public static int size() {
        return list.size();
    }

    public synchronized static int getNormalCount() {
        int count = 0;
        for (AlbumController controller : list) {
            if (controller instanceof NinjaWebView) {
                if (!((NinjaWebView) controller).isIncognito()) {
                    count++;
                }
            } else {
                count++;
            }
        }
        return count;
    }

    public synchronized static int getIncognitoCount() {
        int count = 0;
        for (AlbumController controller : list) {
            if (controller instanceof NinjaWebView && ((NinjaWebView) controller).isIncognito()) {
                count++;
            }
        }
        return count;
    }

    public synchronized static void clear() {
        for (AlbumController albumController : list) {
            if (albumController instanceof NinjaWebView) {
                ((NinjaWebView) albumController).destroy();
            }
        }
        list.clear();
    }
}

File: app/src/main/java/com/petal/browser/browser/BrowserContainer.kt
package com.petal.browser.browser

import com.petal.browser.view.NinjaWebView
import java.util.LinkedList

object BrowserContainer {
    private val list: MutableList<AlbumController> = LinkedList()

    @JvmStatic
    fun get(index: Int): AlbumController {
        return list[index]
    }

    @Synchronized
    @JvmStatic
    fun add(controller: AlbumController) {
        list.add(controller)
    }

    @Synchronized
    @JvmStatic
    fun add(controller: AlbumController, index: Int) {
        list.add(index, controller)
    }

    @Synchronized
    @JvmStatic
    fun remove(controller: AlbumController) {
        if (controller is NinjaWebView) {
            controller.destroy()
        }
        list.remove(controller)
    }

    @JvmStatic
    fun indexOf(controller: AlbumController): Int {
        return list.indexOf(controller)
    }

    @JvmStatic
    fun list(): List<AlbumController> {
        return list
    }

    @JvmStatic
    fun size(): Int {
        return list.size
    }

    @Synchronized
    @JvmStatic
    fun getNormalCount(): Int {
        var count = 0
        for (controller in list) {
            if (controller is NinjaWebView) {
                if (!controller.isIncognito) {
                    count++
                }
            } else {
                count++
            }
        }
        return count
    }

    @Synchronized
    @JvmStatic
    fun getIncognitoCount(): Int {
        var count = 0
        for (controller in list) {
            if (controller is NinjaWebView && controller.isIncognito) {
                count++
            }
        }
        return count
    }

    @Synchronized
    @JvmStatic
    fun clear() {
        for (controller in list) {
            if (controller is NinjaWebView) {
                controller.destroy()
            }
        }
        list.clear()
    }
}
